File: backend/app/schemas/swap.py
from pydantic import BaseModel, validator
from typing import Optional, List
from datetime import datetime
from ..models.swap import SwapStatus
from .skill import SkillBase

class SwapRequestBase(BaseModel):
    responder_id: int
    offered_skill_id: int
    wanted_skill_id: int
    message: Optional[str] = None

class SwapRequestCreate(SwapRequestBase):
    @validator('message')
    def validate_message(cls, v):
        if v and len(v) > 500:
            raise ValueError('Message must be less than 500 characters')
        return v

class SwapRequestUpdate(BaseModel):
    status: SwapStatus
    
class SwapRequestResponse(BaseModel):
    id: int
    requester_id: int
    responder_id: int
    requester_name: str
    responder_name: str
    offered_skill: SkillBase
    wanted_skill: SkillBase
    status: SwapStatus
    message: Optional[str] = None
    created_at: datetime
    has_rating: bool = False
    
    class Config:
        from_attributes = True

class MySwapsResponse(BaseModel):
    pending: List[SwapRequestResponse]
    accepted: List[SwapRequestResponse]
    completed: List[SwapRequestResponse]
    history: List[SwapRequestResponse]

File: frontend/src/types/swap.ts
import { Skill } from './user';

export type SwapStatus = 'pending' | 'accepted' | 'rejected' | 'cancelled' | 'completed';

export interface SwapRequest {
  id: number;
  requester_id: number;
  responder_id: number;
  requester_name: string;
  responder_name: string;
  offered_skill: Skill;
  wanted_skill: Skill;
  status: SwapStatus;
  message?: string;
  created_at: string;
  has_rating: boolean;
}

export interface SwapRequestCreate {
  responder_id: number;
  offered_skill_id: number;
  wanted_skill_id: number;
  message?: string;
}

export interface SwapRequestUpdate {
  status: SwapStatus;
}

export interface MySwaps {
  pending: SwapRequest[];
  accepted: SwapRequest[];
  completed: SwapRequest[];
  history: SwapRequest[];
}

export interface Rating {
  id: number;
  swap_id: number;
  rater_id: number;
  rated_id: number;
  rater_name: string;
  rated_name: string;
  stars: number;
  comment?: string;
  created_at: string;
}

export interface RatingCreate {
  swap_id: number;
  rated_id: number;
  stars: number;
  comment?: string;
}